runtime-sdk/src/modules/rofl/mod.rs

```diff
@@ -426,14 +445,45 @@ impl<Cfg: Config> Module<Cfg> {
         ctx: &C,
         app: &app_id::AppId,
         kind: types::KeyKind,
+        scope: types::KeyScope,
         key_id: &[u8],
+        reg: Option<types::Registration>,
     ) -> Result<keymanager::KeyPair, Error> {
-        let key_id = keymanager::get_key_pair_id([
-            ROFL_DERIVE_KEY_CONTEXT,
-            app.as_ref(),
-            &[kind as u8],
-            key_id,
-        ]);
+        // Build the base key identifier.
+        let kind_id = &[kind as u8];
+        let mut key_id = vec![ROFL_DERIVE_KEY_CONTEXT, app.as_ref(), kind_id, key_id];
+
+        // Append scope for non-global scopes.
+        let scope_id = &[scope as u8];
+        if !scope.is_global() {
+            key_id.push(scope_id);
```

Comment:
Not sure if just appending raw values to `key_id` is a good solution. I think that we should somehow tie this to scope, and not just add raw `u8` or `[u8]` value, like context does. Or change `key_id` so that it's structure is obvious, e.g. key id = kind + optional scope id + optional scope extra data + some optional data in the future.

Reply:
One thing is that we need to retain backwards compatibility for the global scope as this is what is currently deployed. So currently, the structure of the tuple is:
* Context domain separator.
* App ID.
* Encoded kind.
* Key ID.
* Encoded scope.
* Encoded scope-specific domain separation. If there are other scopes in the future, this scope-specific data would all go into this single element.

What did you have in mind that would work without changing the existing keys?

Reply:
We could also introduce a V2 and completely change the structure. But this V2 would only be used for non-global scopes.

Reply:
Currently, if I want to add a new field, e.g. `timeout` or `deadline`, I cannot append it to the list, as scope is optional. So I was thinking if we can instead add a new field (e.g. struct `tags`), to which we can latter add extra data and the byte encoding would be unique, as encoded scope would be tied to something, e.g. to a byte string "scope".

> * V1 context domain separator.
> * App ID.
> * Encoded kind.
> * Key ID.
> * Encoded scope.
> * Encoded scope-specific domain separation. If there are other scopes in the future, this scope-specific data would all go into this single element.

It would be nice if the context separator should have this in the comment.

Reply:
Changed, please take another look.